src/probabilities_estimators/permutation_ordinal/spatial_permutation.jl

```diff
@@ -77,6 +106,29 @@ function SpatialSymbolicPermutation(
     SpatialSymbolicPermutation{D, p, typeof(valid)}(stencil, copy(stencil), arraysize, valid)
 end
 
+function SpatialSymbolicPermutation(
+    stencil::NTuple{2, NTuple{D, T}}, x::AbstractArray, p::Bool = true
+    ) where {D, T}
+    # get extent and lag from stencil
+    extent, lag = stencil
+    # generate a D-dimensional stencil
+    # start by generating a list of iterators for each dimension
+    iters = [0:lag[i]:extent[i]-1 for i in 1:D]
+    # then generate the stencil. We use an iterator product that we basically only reshape after that
+    stencil = CartesianIndex.(vcat(collect(Iterators.product(iters...))...))
+    SpatialSymbolicPermutation(stencil, x, p)
+end
+
+function SpatialSymbolicPermutation(
```

Comment:
Instead of repeating `SpatialSymbolicPermutation` three times, it makes for simpler design, and more future proof, to define an internal function `stencil_to_offsets` that dispatches on stencil type and always gives the vector of cartesians. in the case where stencil is already vector cartesisans the functio nreturns its argument.

Reply:
Agreed.